File: typescript/examples/orchestration.mjs
import { setTimeout as delay } from "node:timers/promises";
import { pathToFileURL } from "node:url";
import path from "node:path";
import { Admin, Pool, Queue, Worker } from "@stablemates/workhorse";

export function createOrderWorkers(queue) {
  const parentWorker = new Worker(queue, {
    queue: "orders",
    workerId: "orders-orchestrator",
  }).handle("order.process", async (payload, context) => {
    const children = await context.runChildrenAll([
      { name: "invoice", type: "invoice.create", payload },
      { name: "receipt", type: "receipt.send", payload },
    ]);
    return { children };
  });

  const childWorker = new Worker(queue, {
    queue: "orders",
    workerId: "orders-steps",
    concurrency: 2,
  })
    .handle("invoice.create", async (payload) => ({ completed: true, payload }))
    .handle("receipt.send", async (payload) => ({ completed: true, payload }));

  return { parentWorker, childWorker };
}

export async function runOrchestrationExample(databaseUrl) {
  const pool = new Pool({ connectionString: databaseUrl, max: 4 });
  try {
    const queue = new Queue(pool, "orders");
    const admin = new Admin(pool);
    const workers = createOrderWorkers(queue);
    const jobId = await queue.enqueue("order.process", { orderId: "order-42" });

    for (let pass = 0; pass < 100; pass += 1) {
      await workers.parentWorker.runOnce();
      await workers.childWorker.runOnce();
      const job = await admin.getJob(jobId);
      if (job?.state === "succeeded") return { jobId, result: job.result };
      if (job?.state === "failed" || job?.state === "canceled") {
        throw new Error(`Order orchestration finished in ${job.state} state`);
      }
      await delay(20);
    }
    throw new Error(`Order orchestration did not finish ${jobId}`);
  } finally {
    await pool.end();
  }
}

const invokedPath = process.argv[1] ? pathToFileURL(path.resolve(process.argv[1])).href : null;
if (import.meta.url === invokedPath) {
  const databaseUrl = process.env.DATABASE_URL;
  if (!databaseUrl) throw new Error("DATABASE_URL is required");
  process.stdout.write(`${JSON.stringify(await runOrchestrationExample(databaseUrl))}\n`);
}

File: rust/examples/orchestration.rs
//! Runs an order handler that fans out two child tasks, then waits for a signal and a human decision.
use serde_json::{json, Value};
use workhorse::deadpool_postgres::tokio_postgres::NoTls;
use workhorse::deadpool_postgres::{Manager, Pool};
use workhorse::{
    run_worker_process, ChildTaskRequest, HandlerContext, HandlerError, Worker, WorkerOptions,
};

async fn process_order(payload: Value, context: HandlerContext) -> Result<Value, HandlerError> {
    let children = context
        .run_children_all(vec![
            ChildTaskRequest::new("invoice", "invoice.create", &payload)?,
            ChildTaskRequest::new("receipt", "receipt.send", &payload)?,
        ])
        .await?;
    let approval: Value = context.wait_for_signal("approval", None).await?.payload;
    let review = json!({ "payload": payload, "approval": approval });
    let decision: Value = context.wait_for_human("review", &review, None).await?.result;
    Ok(json!({ "children": children, "approval": approval, "decision": decision }))
}

async fn complete_child(payload: Value, _context: HandlerContext) -> Result<Value, HandlerError> {
    Ok(json!({ "completed": true, "payload": payload }))
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let url = std::env::var("WORKHORSE_DATABASE_URL")?;
    let pool = Pool::builder(Manager::new(url.parse()?, NoTls)).max_size(10).build()?;

    let worker =
        Worker::new(pool, WorkerOptions { queues: vec!["orders".into()], ..Default::default() })?;
    worker.handle("order.process", process_order);
    worker.handle("invoice.create", complete_child);
    worker.handle("receipt.send", complete_child);
    run_worker_process(&worker).await?;
    Ok(())
}
